### ISSUE TITLE
[openssl] build failure on alpine linux due to missing headers

### ISSUE THREAD

**Issue by mprather** [microsoft/vcpkg#23889]
**Host Environment**
- OS: Alpine Linux 3.15.3
- Compiler: g++ (Alpine 10.3.1_git20211027) 10.3.1 20211027

**To Reproduce**

1. Install minimum set of system packages needed to run vcpkg on an alpine system.
2. ensure system package perl is added
> apk add perl

3. git clone vckpg
4. set env variable VCPKG_FORCE_SYSTEM_BINARIES to 1
5. run bootstrap-vcpkg.sh (it pulls down the pre-build muslc binary of vcpkg)
6. Attempt to install openssl

  > ./vcpkg install openssl

vcpkg reports:
```
-- Building x64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):
    Command failed: /usr/bin/cmake --build . --config Debug --target install -- -v -j9
    Working Directory: /src/vcpkg/buildtrees/openssl/x64-linux-dbg
    See logs for more information:
      /src/vcpkg/buildtrees/openssl/install-x64-linux-dbg-out.log
      /src/vcpkg/buildtrees/openssl/install-x64-linux-dbg-err.log

Call Stack (most recent call first):
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:111 (vcpkg_execute_build_process)
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:41 (vcpkg_cmake_build)
  ports/openssl/unix/portfile.cmake:27 (vcpkg_cmake_install)
  ports/openssl/portfile.cmake:24 (include)
  scripts/ports.cmake:145 (include)


Error: Building package openssl:x64-linux failed with: BUILD_FAILED
```

**Failure logs**
Output log shows root cause...
```
/usr/bin/cc  -I. -Iinclude -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -fPIC -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/src/vcpkg/packages/openssl_x64-linux/debug/lib/engines-1.1\"" -DNDEBUG  -MMD -MF crypto/modes/cfb128.d.tmp -MT crypto/modes/cfb128.o -c -o crypto/modes/cfb128.o crypto/modes/cfb128.c
crypto/mem_sec.c:33:13: fatal error: linux/mman.h: No such file or directory
   33 | #   include <linux/mman.h>
      |             ^~~~~~~~~~~~~~
compilation terminated.
```

**Additional context**
This problem can be resolved by adding a system package.

> apk add linux-headers

**Expectation**
The port should tell the user what system packages are missing, if any, in order to build the port.  The openssl was able to report that perl was needed - seems natural to also inform about this requirement.

**Comment by StarGate-One** [comment]
Has the [Linux Developer Tools](https://github.com/microsoft/vcpkg#installing-linux-developer-tools) been installed, as described in the vcpkg readme documentation?

**Comment by FrankXie05** [comment]
@mprather Thanks for posting this issue, This issue seems to be discussed upstream, you can try this https://github.com/openssl/openssl/issues/7207#issuecomment-442526916,

**Comment by mprather** [comment]
Note that the page does not outline Alpine packages.

In #21218, I have documented a typical alpine setup.
* newest cmake from 3.15 repo
* and the following packages: build-base curl git g++ make ninja pkgconfig tar unzip zip

Build-essentials does not exist in alpine.  For the most part, the list above has been working for the better part of a year. This is only 1 of 2 packages in that time that we've seen that require additional system packages.  

As noted above, if the linux-headers package is added, the problem is resolved. I would hope that if cmake scripts can safely call out that compilation depends on perl when it is not present, then they should also call out the requirement of the library-headers package when a) on alpine and b) it is not present.

**Comment by StarGate-One** [comment]
@mprather I apologize, I did not realize this was related to an earlier issue - most of my `*ix` experience is with AIX, HP-UX and Solaris although I have looked around a little with Red Hat, OpenSUSE and Arch linux.

**Comment by mprather** [comment]
@StarGate-One no worries. boundaries are quite fuzzy in two areas. the first, using vcpkg in alpine is slowly becoming a supported config. In the past quarter or so, the vcpkg has made really good strides in this area.  For example, the bootstrap now tells you it needs "build-base cmake ninja zip unzip curl git" if it doesn't find a prereq. the second fuzzy area is whether a port is doing more to setup its config for use in vcpkg. this quite honestly is a more opaque scenario simply because I have not delved into port dev/management. should the port author add a check? or should a port author depend on the upstream config to be smarter?  I really don't have a good idea. I just wanted to document this problem to help add precedence for alpine users.

**Comment by FrankXie05** [comment]
Fixed by #24004
